openzwavemqtt/util/mqtt_client.py

```diff
@@ -0,0 +1,246 @@
+        """
+        params: dict = {"timeout": timeout}
+        if properties:
+            params["properties"] = properties
+
+        await self.asyncio_client.unsubscribe(topic, **params)
+
+    def send_message(self, topic: str, payload: Union[str, dict]) -> None:
+        """Send a message from the manager options."""
+        to_publish = (topic, json.dumps(payload))
+        self.publish_queue.put_nowait(to_publish)
+
+    async def handle_publish(self) -> None:
+        """Publish messages as they are put on the queue."""
+        while True:
+            to_publish: tuple = await self.publish_queue.get()
+            await self.publish(*to_publish)
+            self.publish_queue.task_done()
+
+    async def start_client(self, manager: OZWManager) -> None:
+        """Start the client with the manager."""
+        self.client_task = asyncio.current_task()
+        # Reconnect automatically until the client is stopped.
+        while True:
+            try:
+                await self._subscribe_manager(manager)
+            except MqttError as err:
+                self.reconnect_interval = min(self.reconnect_interval * 2, 900)
+                LOGGER.error(
+                    "MQTT error: %s. Reconnecting in %s seconds",
+                    err,
+                    self.reconnect_interval,
+                )
+                await asyncio.sleep(self.reconnect_interval)
+                self.create_client()  # reset connect/reconnect futures
+
+    async def _subscribe_manager(self, manager: OZWManager) -> None:
+        """Connect and subscribe to manager topics."""
+        async with AsyncExitStack() as stack:
+            # Keep track of the asyncio tasks that we create, so that
+            # we can cancel them on exit.
+            tasks: Set[asyncio.Task] = set()
+            stack.push_async_callback(cancel_tasks, tasks)
+
+            # Connect to the MQTT broker.
+            await stack.enter_async_context(self.asyncio_client)
+            # Reset the reconnect interval after successful connection.
+            self.reconnect_interval = 1
+
+            publish_task = asyncio.create_task(self.handle_publish())
+            tasks.add(publish_task)
+
+            # Messages that doesn't match a filter will get logged and handled here.
+            messages = await stack.enter_async_context(
+                self.asyncio_client.unfiltered_messages()
+            )
+
+            messages_task = asyncio.create_task(
+                handle_messages(messages, manager.receive_message)
+            )
+            tasks.add(messages_task)
+
+            # Note that we subscribe *after* starting the message loggers.
+            # Otherwise, we may miss retained messages.
+            topic = f"{manager.options.topic_prefix}#"
+            await self.subscribe(topic)
+
+            # Wait for everything to complete (or fail due to, e.g., network errors).
+            await asyncio.gather(*tasks)
```

Comment:
I think that instead of a gather, why don't we do something like:

```python
publish_task = asyncio.create_task(…)
try:
    async for message in messages:
        ...
finally:
    publish_task.cancel()
```

Would make it more clear what's going on.

Reply:
You can also remove the `cancel_tasks` function then.

Reply:
No, we can't do that since we want to subscribe after starting the message handler.

Reply:
At least, I'm not sure that's safe since we await the subscribe so we yield control to the loop and potentially retained messages could be coming in before we get control back?